File: ts/src/webex.ts
export const WEBEX_API_BASE = 'https://webexapis.com/v1';
export let WEBEX_MESSAGES_URL = `${WEBEX_API_BASE}/messages`;
export let WEBEX_ATTACHMENT_ACTIONS_URL = `${WEBEX_API_BASE}/attachment/actions`;
export const ADAPTIVE_CARD_CONTENT_TYPE = 'application/vnd.microsoft.card.adaptive';

const MAX_RETRIES = 3;

let sleepFn = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

function sleep(ms: number): Promise<void> {
  return sleepFn(ms);
}

interface CardAttachment {
  contentType: string;
  content: unknown;
}

interface SendMessagePayload {
  roomId?: string;
  toPersonEmail?: string;
  text: string;
  attachments?: CardAttachment[];
}

export async function sendMessage(
  token: string,
  toEmail: string,
  text: string,
  fetchFn: typeof fetch = globalThis.fetch,
  roomId = '',
): Promise<Record<string, unknown>> {
  const payload: SendMessagePayload = roomId
    ? { roomId, text }
    : { toPersonEmail: toEmail, text };
  return postMessage(token, payload, fetchFn);
}

export async function sendCard(
  token: string,
  toEmail: string,
  text: string,
  card: unknown,
  fetchFn: typeof fetch = globalThis.fetch,
  roomId = '',
): Promise<Record<string, unknown>> {
  const payload: SendMessagePayload = roomId
    ? {
        roomId,
        text,
        attachments: [{ contentType: ADAPTIVE_CARD_CONTENT_TYPE, content: card }],
      }
    : {
        toPersonEmail: toEmail,
        text,
        attachments: [{ contentType: ADAPTIVE_CARD_CONTENT_TYPE, content: card }],
      };
  return postMessage(token, payload, fetchFn);
}

async function postMessage(
  token: string,
  payload: SendMessagePayload,
  fetchFn: typeof fetch,
): Promise<Record<string, unknown>> {
  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    const resp = await fetchFn(WEBEX_MESSAGES_URL, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });

    if (resp.status === 429) {
      if (attempt < MAX_RETRIES) {
        const retryAfter = resp.headers.get('Retry-After') || '1';
        const delaySecs = Math.min(parseInt(retryAfter, 10) || 1, 300);
        await sleep(delaySecs * 1000);
        continue;
      }
    }

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`HTTP ${resp.status}: ${body}`);
    }
    return (await resp.json()) as Record<string, unknown>;
  }

  // This should be unreachable, but TypeScript needs it
  throw new Error('Unexpected state in postMessage retry loop');
}

export async function getMessage(
  token: string,
  messageId: string,
  fetchFn: typeof fetch = globalThis.fetch,
): Promise<Record<string, unknown>> {
  return getJSON(token, `${WEBEX_MESSAGES_URL}/${messageId}`, fetchFn);
}

export async function getAttachmentAction(
  token: string,
  actionId: string,
  fetchFn: typeof fetch = globalThis.fetch,
): Promise<Record<string, unknown>> {
  return getJSON(token, `${WEBEX_ATTACHMENT_ACTIONS_URL}/${actionId}`, fetchFn);
}

async function getJSON(
  token: string,
  url: string,
  fetchFn: typeof fetch,
): Promise<Record<string, unknown>> {
  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    const resp = await fetchFn(url, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });

    if (resp.status === 429) {
      if (attempt < MAX_RETRIES) {
        const retryAfter = resp.headers.get('Retry-After') || '1';
        const delaySecs = Math.min(parseInt(retryAfter, 10) || 1, 300);
        await sleep(delaySecs * 1000);
        continue;
      }
    }

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`HTTP ${resp.status}: ${body}`);
    }
    return (await resp.json()) as Record<string, unknown>;
  }

  // This should be unreachable, but TypeScript needs it
  throw new Error('Unexpected state in getJSON retry loop');
}

export function extractCards(message: Record<string, unknown>): unknown[] {
  const attachments = message.attachments;
  if (!Array.isArray(attachments)) return [];
  const cards: unknown[] = [];
  for (const att of attachments) {
    if (att && typeof att === 'object' && 'contentType' in att && 'content' in att) {
      if ((att as Record<string, unknown>).contentType === ADAPTIVE_CARD_CONTENT_TYPE) {
        cards.push((att as Record<string, unknown>).content);
      }
    }
  }
  return cards;
}

/** Override message URL for testing */
export function _setMessagesUrl(url: string): void {
  WEBEX_MESSAGES_URL = url;
}

export function _setAttachmentActionsUrl(url: string): void {
  WEBEX_ATTACHMENT_ACTIONS_URL = url;
}

/** Override sleep function for testing */
export function _setSleepFn(fn: (ms: number) => Promise<void>): void {
  sleepFn = fn;
}

File: rust/src/webex.rs
use reqwest::Client;
use serde::Serialize;
use serde_json::Value;
use std::sync::RwLock;
use std::time::Duration;
use tokio::time::sleep;

pub const WEBEX_API_BASE: &str = "https://webexapis.com/v1";
pub const ADAPTIVE_CARD_CONTENT_TYPE: &str = "application/vnd.microsoft.card.adaptive";
const MAX_RETRIES: u32 = 3;

static MESSAGES_URL_OVERRIDE: RwLock<Option<String>> = RwLock::new(None);
static ATTACHMENT_ACTIONS_URL_OVERRIDE: RwLock<Option<String>> = RwLock::new(None);

fn messages_url() -> String {
    if let Ok(guard) = MESSAGES_URL_OVERRIDE.read() {
        if let Some(url) = guard.as_ref() {
            return url.clone();
        }
    }
    format!("{}/messages", WEBEX_API_BASE)
}

fn attachment_actions_url() -> String {
    if let Ok(guard) = ATTACHMENT_ACTIONS_URL_OVERRIDE.read() {
        if let Some(url) = guard.as_ref() {
            return url.clone();
        }
    }
    format!("{}/attachment/actions", WEBEX_API_BASE)
}

/// Override messages URL for testing. Pass None to reset.
pub fn _set_messages_url(url: Option<String>) {
    *MESSAGES_URL_OVERRIDE.write().unwrap() = url;
}

/// Override attachment actions URL for testing. Pass None to reset.
pub fn _set_attachment_actions_url(url: Option<String>) {
    *ATTACHMENT_ACTIONS_URL_OVERRIDE.write().unwrap() = url;
}

#[derive(Serialize)]
struct SendMessagePayload {
    #[serde(rename = "toPersonEmail")]
    to_person_email: String,
    text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    attachments: Option<Vec<CardAttachment>>,
}

#[derive(Serialize)]
struct CardAttachment {
    #[serde(rename = "contentType")]
    content_type: String,
    content: Value,
}

pub async fn send_message(
    token: &str,
    to_email: &str,
    text: &str,
    client: &Client,
) -> Result<Value, String> {
    let payload = SendMessagePayload {
        to_person_email: to_email.to_string(),
        text: text.to_string(),
        attachments: None,
    };
    post_message(token, &payload, client).await
}

pub async fn send_card(
    token: &str,
    to_email: &str,
    text: &str,
    card: &Value,
    client: &Client,
) -> Result<Value, String> {
    let payload = SendMessagePayload {
        to_person_email: to_email.to_string(),
        text: text.to_string(),
        attachments: Some(vec![CardAttachment {
            content_type: ADAPTIVE_CARD_CONTENT_TYPE.to_string(),
            content: card.clone(),
        }]),
    };
    post_message(token, &payload, client).await
}

async fn post_message(
    token: &str,
    payload: &SendMessagePayload,
    client: &Client,
) -> Result<Value, String> {
    let payload_json = serde_json::to_string(payload)
        .map_err(|e| format!("serialize payload: {}", e))?;

    let mut attempt = 0;
    loop {
        let resp = client
            .post(messages_url())
            .header("Authorization", format!("Bearer {}", token))
            .header("Content-Type", "application/json")
            .body(payload_json.clone())
            .send()
            .await
            .map_err(|e| format!("send message: {}", e))?;

        let status = resp.status();

        if status.as_u16() == 429 && attempt < MAX_RETRIES {
            let retry_after = resp
                .headers()
                .get("Retry-After")
                .and_then(|h| h.to_str().ok())
                .and_then(|s| s.parse::<u64>().ok())
                .unwrap_or(1)
                .min(300);

            sleep(Duration::from_secs(retry_after)).await;
            attempt += 1;
            continue;
        }

        let body = resp.text().await.map_err(|e| format!("read response: {}", e))?;

        if !status.is_success() {
            return Err(format!("HTTP {}: {}", status.as_u16(), body));
        }
        return serde_json::from_str(&body).map_err(|e| format!("parse response: {}", e));
    }
}

pub async fn get_message(
    token: &str,
    message_id: &str,
    client: &Client,
) -> Result<Value, String> {
    get_json(token, &format!("{}/{}", messages_url(), message_id), client).await
}

pub async fn get_attachment_action(
    token: &str,
    action_id: &str,
    client: &Client,
) -> Result<Value, String> {
    get_json(
        token,
        &format!("{}/{}", attachment_actions_url(), action_id),
        client,
    )
    .await
}

async fn get_json(token: &str, url: &str, client: &Client) -> Result<Value, String> {
    let mut attempt = 0;
    loop {
        let resp = client
            .get(url)
            .header("Authorization", format!("Bearer {}", token))
            .header("Content-Type", "application/json")
            .send()
            .await
            .map_err(|e| format!("GET {}: {}", url, e))?;

        let status = resp.status();

        if status.as_u16() == 429 && attempt < MAX_RETRIES {
            let retry_after = resp
                .headers()
                .get("Retry-After")
                .and_then(|h| h.to_str().ok())
                .and_then(|s| s.parse::<u64>().ok())
                .unwrap_or(1)
                .min(300);

            sleep(Duration::from_secs(retry_after)).await;
            attempt += 1;
            continue;
        }

        let body = resp.text().await.map_err(|e| format!("read response: {}", e))?;

        if !status.is_success() {
            return Err(format!("HTTP {}: {}", status.as_u16(), body));
        }
        return serde_json::from_str(&body).map_err(|e| format!("parse response: {}", e));
    }
}

pub fn extract_cards(message: &Value) -> Vec<Value> {
    let attachments = match message.get("attachments").and_then(|a| a.as_array()) {
        Some(arr) => arr,
        None => return vec![],
    };
    attachments
        .iter()
        .filter_map(|att| {
            let ct = att.get("contentType")?.as_str()?;
            if ct == ADAPTIVE_CARD_CONTENT_TYPE {
                att.get("content").cloned()
            } else {
                None
            }
        })
        .collect()
}
